File: Python/ESTRUTURAS E BIBLIOTECA/1259 Pares e Ímpares.py
# SUBPROGRAMA
def quicksort(arr, start, end):
    if start < end:
        if arr[0] % 2 == 0:  # if arr[0] == even number
            pivotPosition = partitionAscending(arr, start, end)
            quicksort(arr, start, pivotPosition-1)
            quicksort(arr, pivotPosition+1, end)
        else:  # if arr[0] == odd number
            pivotPosition = partitionDescending(arr, start, end)
            quicksort(arr, start, pivotPosition-1)
            quicksort(arr, pivotPosition+1, end)
    return None

def partitionAscending(arr, start, end):
    pivot = arr[end]
    i = start-1
    for j in range(start, end):
        if arr[j] <= pivot:
            i += 1
            arr[j], arr[i] = arr[i], arr[j]
    i += 1
    arr[i], arr[end] = arr[end], arr[i]  # pivot = arr[end]
    return i

def partitionDescending(arr, start, end):
    pivot = arr[end]
    i = start-1
    for j in range(start, end):
        if arr[j] >= pivot:
            i += 1
            arr[j], arr[i] = arr[i], arr[j]
    i += 1
    arr[i], arr[end] = arr[end], arr[i]  # pivot = arr[end]
    return i


# PROGRAMA PRINCIPAL
lengh = int(input())
evenNumbers = list()
oddNumbers = list()
for i in range(0, lengh):
    number = int(input())
    if number % 2 == 0:
        evenNumbers += [number]
    else:
        oddNumbers += [number]
quicksort(evenNumbers, 0, len(evenNumbers)-1)
quicksort(oddNumbers, 0, len(oddNumbers)-1)
for i in range(0, len(evenNumbers)):
    print("%d" % evenNumbers[i])
for i in range(0, len(oddNumbers)):
    print("%d" % oddNumbers[i])

File: C/Estruturas e Bibliotecas/1259 Pares e Ímpares.c
#include <stdio.h>
#include <stdlib.h>

int quicksort(int *arr, int start, int end);
int partitionAscending(int *arr, int start, int end);
int partitionDescending(int *arr, int start, int end);


int main(){
    int len;
    scanf("%d", &len);
    int *evenNumbers, *oddNumbers, number, evenCount=0, oddCount=0;
    evenNumbers = (int*) malloc(len*sizeof(int));
    oddNumbers = (int*) malloc(len*sizeof(int));
    while (len != 0){
        scanf("%d", &number);
        if (number % 2 == 0){
            evenNumbers[evenCount] = number;
            evenCount += 1;
        }
        else{
            oddNumbers[oddCount] = number;
            oddCount += 1;
        }
        len -= 1;
    } 

    // evenCount-1 and oddCount-1 are the final positions of the array
    quicksort(evenNumbers, 0, evenCount-1); 
    quicksort(oddNumbers, 0, oddCount-1);

    for (int i=0; i < evenCount; i++)
        printf("%d\n", evenNumbers[i]);
    for (int i=0; i < oddCount; i++)
        printf("%d\n", oddNumbers[i]);
    free(evenNumbers);
    free(oddNumbers);
}

int quicksort(int *arr, int start, int end){
    
    if (start < end)
    {
        if (arr[0] % 2 == 0) // if arr[0] == even number
        {
            int pivotPosition = partitionAscending(arr, start, end);
            quicksort(arr, start, pivotPosition-1);
            quicksort(arr, pivotPosition+1, end);
        }
        else // if arr[0] == odd number
        {
            int pivotPosition = partitionDescending(arr, start, end);
            quicksort(arr, start, pivotPosition-1);
            quicksort(arr, pivotPosition+1, end);
        }
    }
    return 0;
}

int partitionAscending(int *arr, int start, int end)
{
    int pivot = arr[end], i=start-1, temp;
    for (int j=start; j < end; j++)
    {
        if (arr[j] <= pivot)
        {
            i += 1;
            temp = arr[j];
            arr[j] = arr[i];
            arr[i] = temp;
        }
    }
    i += 1;
    temp = arr[i];
    arr[i] = arr[end]; // pivot == arr[end]
    arr[end] = temp;
    return i;
}


int partitionDescending(int *arr, int start, int end)
{
    int pivot = arr[end], i=start-1, temp;
    for (int j=start; j < end; j++)
    {
        if (arr[j] >= pivot)
        {
            i += 1;
            temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
    i += 1;
    temp = arr[i];
    arr[i] = arr[end]; //pivot == arr[end]
    arr[end] = temp;
    return i;
}
